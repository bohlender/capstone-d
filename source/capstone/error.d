module capstone.error;

import std.conv;
import std.exception: enforce, assertNotThrown;
import std.format;
import std.string;

import capstone.internal.api: cs_strerror;

enum ErrorCode {
    Ok = 0,
    OutOfMemory,
    UnsupportedArchitecture,
    InvalidHandle,
    InvalidCshArgument,
    InvalidMode,
    InvalidOption,
    UnavailableInstructionDetail,
    UninitializedDynamicMemoryManagement,
    UnsupportedVersion,
    IrrelevantDataAccessInDietEngine,
    IrrelevantDataAccessInSkipdataMode,
    UnsupportedATnTSyntax,
    UnsupportedIntelSyntax
}

// TODO: Proper error handling
void checkErrorCode(in ErrorCode errno){
    enforce(errno == 0, "Error %s: %s".format(errno.to!int, cs_strerror(errno.to!int).fromStringz));
}

void checkErrorCode(in int errno)
in{
    assertNotThrown!ConvException(errno.to!ErrorCode);
}
body{
    errno.to!ErrorCode.checkErrorCode;
}