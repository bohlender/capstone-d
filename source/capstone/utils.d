module capstone.utils;

import std.meta: AliasSeq, staticMap;
import std.format: format;
import std.exception: enforce;

private enum bool distinctFieldNames(names...) = __traits(compiles,{
    static foreach (name; names)
        static if (is(typeof(name) : string))
            mixin("enum int" ~ name ~ " = 0;");
});

private enum bool distinctTypeNames(names...) = __traits(compiles,{
    static foreach (name; names)
        static if (is(name))
            mixin("enum int" ~ name.mangleof ~ " = 0;");
});

unittest{
	assert(distinctFieldNames!("asdf","asd",int,"asdg"));
	assert(!distinctFieldNames!("asdf","asd","asdf"));
	assert(distinctTypeNames!(int,long,string,"string",uint));
	assert(!distinctTypeNames!(int,long,uint,int));
}

template TaggedUnion(Specs...) if (distinctFieldNames!Specs && distinctTypeNames!Specs){
	template FieldSpec(T, string s){
		alias Type = T;
		alias name = s;
	}
	
	template parseSpecs(Specs...){
		static if(Specs.length==0){
			alias parseSpecs = AliasSeq!();
		}else static if(is(Specs[0])){
			alias parseSpecs = AliasSeq!(FieldSpec!(Specs[0..2]), parseSpecs!(Specs[2..$]));
		}else{
			static assert(0, "Attempted to instantiate TaggedUnion with an invalid argument: " ~ Specs[0].stringof);
		}
		
	}

	alias fieldSpecs = parseSpecs!Specs;
	alias extractType(alias spec) = spec.Type;
	alias extractName(alias spec) = spec.name;
	alias Types = staticMap!(extractType, fieldSpecs);
	alias names = staticMap!(extractName, fieldSpecs);

	struct TaggedUnion{
		import std.variant: Algebraic;
		private Algebraic!Types wrapped;

		// Generate constructors
		static foreach(T; Types){
			this(T val){wrapped = val;}
		}

		// Generate getters
		static foreach(spec; fieldSpecs){
			mixin("@property auto " ~ spec.name ~"() const {
				enforce(wrapped.type == typeid(spec.Type), \"Trying to read a '\" ~ spec.Type.stringof ~ \"' but a '\" ~ wrapped.type.toString ~ \"' is stored\");
				return wrapped.get!(spec.Type);
			}");
		}

		// Generate setters
		static foreach(i, spec; fieldSpecs){
			mixin("@property auto " ~ spec.name ~ "(spec.Type val){wrapped = val;}");
		}
	}
}