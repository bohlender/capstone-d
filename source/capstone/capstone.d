/// Object-oriented wrapper of Capstone disassembly engine
module capstone.capstone;

import std.typecons: Tuple, BitFlags, Yes, Nullable;
import std.format: format;
import std.conv: to;
import std.array: array, appender;
import std.range: isInputRange, enumerate, front;
import std.algorithm: canFind;
import std.traits: EnumMembers;

import capstone.api;
import capstone.instruction;
import capstone.internal;
import capstone.error;

/** Encapsulates an instance of the Capstone dissassembly engine

This class encapsulates the core functionality of the Capstone disassembly engine, providing
access to runtime options for
$(UL
    $(LI changing the `Mode` of interpretation)
    $(LI changing the `Syntax` of the disassembly)
    $(LI choosing whether `Instruction`'s should be disassembled in detail, i.e. filling `Instruction.detail`)
    $(LI defining manual handling of broken instructions through the $(LINK2 http://www.capstone-engine.org/skipdata.html, SKIPDATA) mode of operation (optionally via a `Callback`))
)

Note that, since the architecture is chosen at runtime, this base class only provides access to the architecture-indepentent aspects,
e.g. disasm returns `Instruction`s instead of `X86Instruction`s.
However, if necessary, it can be casted to the architecture-specific variant, such as `CapstoneX86`.
*/
abstract class Capstone{
    package {
        alias Handle = size_t;
        Handle handle;    
        
        ModeFlags _mode;
        Syntax _syntax;
        bool _detail;
        bool _skipData;

        string mnemonic;
        Callback callback;

        string[int] customMnemonics;
    }
    const Arch arch; /// The architecture this Capstone instance is set up for

    /** Constructs an instance of the disassembly engine

    Params:
        arch = The architecture the engine will be created for
        modeFlags = A combination of flags to further specify how bytes will be interpreted, e.g. in little-endian.
    */
    private this(in Arch arch, in ModeFlags modeFlags){
        const libVer = versionOfLibrary;
        const bindVer = versionOfBindings;
        if(libVer != bindVer)
            throw new CapstoneException("API version mismatch between library (%s) and bindings (%s)".format(libVer, bindVer), ErrorCode.UnsupportedVersion);

        // Create Capstone engine instance
        this.arch = arch;
        this._mode = modeFlags;
        cs_open(arch, modeFlags.to!uint, &handle).checkErrno;
    }

    ~this(){
        if(handle)
           cs_close(&handle).checkErrno;
    }
    
    /// Gets the mode of interpretation
    @property auto mode() const {return _mode;}
    /// Sets the mode of interpretation
    @property void mode(in ModeFlags modeFlags){
        _mode = modeFlags;
        cs_option(handle, cs_opt_type.CS_OPT_MODE, modeFlags.to!uint).checkErrno;
    }

    /// Gets the disassembly syntax variant
    @property auto syntax() const {return _syntax;}
    /// Sets the disassembly syntax variant
    @property void syntax(in Syntax option){
        _syntax = option;
        cs_option(handle, cs_opt_type.CS_OPT_SYNTAX, option).checkErrno;
    }

    /// Indicates whether instructions will be disassembled in detail
    @property auto detail() const {return _detail;}
    /// Sets whether instructions will be disassembled in detail
    @property void detail(in bool enable){
        _detail = enable;
        auto option = enable ? cs_opt_value.CS_OPT_ON : cs_opt_value.CS_OPT_OFF;
        cs_option(handle, cs_opt_type.CS_OPT_DETAIL, option).checkErrno;
    }

    /// Indicates whether SKIPDATA mode of operation is in use
    @property auto skipData() const {return _skipData;}
    /// Sets whether to use SKIPDATA mode of operation
    @property void skipData(in bool enable){
        _skipData = enable;
        auto option = enable ? cs_opt_value.CS_OPT_ON : cs_opt_value.CS_OPT_OFF;
        cs_option(handle, cs_opt_type.CS_OPT_SKIPDATA, option).checkErrno;
    }

    /** Customises behaviour in SKIPDATA mode of operation
     
    By default, disassembling will stop when it encounters a broken instruction.
    Most of the time, the reason is that this is data mixed inside the input.

    When in SKIPDATA mode, some (unknown) amount of data until the next interpretable instruction will be skipped.
    Capstone considers the skipped data a special instruction with ID 0x00 and a `mnemonic` that defaults to `".byte"`.
    The operand string is a hex-code of the sequence of bytes it skipped.

    By default, for each iteration, Capstone skips 1 byte on X86 architecture, 2 bytes on Thumb mode on Arm
    architecture, and 4 bytes for the rest. The reason while Capstone skips 1 byte on X86 is that X86 puts no
    restriction on instruction alignment, but other architectures enforce some requirements on this aspect.

    To customise how many bytes to skip when encountering data, a `Callback` delegate can optonally be setup
    to return the corresponding number.

    Params:
        mnemonic = The mnemonic to use for representing skipped data
        callback = The optional callback to use for handling bytes that cannot be interpreted as an instruction.
    
    Example:
    ---
    // Custom data that can be referred to in a callback delegate
    struct CallbackData{
        int bytesToSkip;
    }
    auto myData = CallbackData(1);

    size_t myCallback(in ubyte[] code, size_t offset) {
        return myData.bytesToSkip++; // Always skip one more byte when encountering data
    }

    cs.skipData = true;                     // Enable skipdata mode
    cs.setupSkipdata("db", &myCallback);    // Use custom callback, and "db" as custom mnemonic for data
    ---
    */
    void setupSkipdata(in string mnemonic = ".byte", Callback callback = null){
        if(!mnemonic)
            throw new CapstoneException("Invalid mnemonic", ErrorCode.InvalidOption);
        this.mnemonic = mnemonic;
        this.callback = callback;
        
        auto setup = cs_opt_skipdata(this.mnemonic.ptr, this.callback ? &cCallback : null, &this.callback);
        cs_option(handle, cs_opt_type.CS_OPT_SKIPDATA_SETUP, cast(size_t)&setup).checkErrno;
    }

    /// Defines a custom mnemonic for a specified instruction id
    private void customMnemonic(in int id, in string mnem = null) {
        auto optMnem = cs_opt_mnem(id, null);
        if(mnem != null){
            const v = (customMnemonics[id] = mnem);
            optMnem.mnemonic = v.ptr;
        }else
            customMnemonics.remove(id);
        cs_option(handle, cs_opt_type.CS_OPT_MNEMONIC, cast(size_t)&optMnem).checkErrno;
    }

    /** Disassemble binary code, given the code buffer, start address and number of instructions to be decoded
    
    Params:
        code    = Buffer containing raw binary code to be disassembled
        address = Address of the first instruction in given raw code buffer
        count   = Number of instructions to be disassembled, or 0 to get all of them
    Returns: The successfully disassembled instructions

    Example:
    ---
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32)); // Initialise x86 32bit engine
    auto res = cs.disasm(CODE, 0x1000);               // Disassemble, offsetting addresses by 0x1000
    assert("%s %s".format(res[0].mnemonic, res[0].opStr) == "lea ecx, dword ptr [edx + esi + 8]");
    assert("%s %s".format(res[1].mnemonic, res[1].opStr) == "add eax, ebx");
    assert("%s %s".format(res[2].mnemonic, res[2].opStr) == "add esi, 0x1234");
    ---
    */
    abstract const(Instruction)[] disasm(in ubyte[] code, in ulong address, in size_t count = 0) const;

    /** Provides a range to iteratively disassemble binary code - one instruction at a time

    Fast API to disassemble binary code, given the code buffer and start address.
    Provides access to only one disassembled instruction at a time, resulting in a smaller memory footprint.
    Params:
        code    = Buffer containing raw binary code to be disassembled
        address = Address of the first instruction in given raw code buffer
    Returns: An input range over the disassembled instructions
    Example:
    ---
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32)); // Initialise x86 32bit engine
    auto range = cs.disasmIter(CODE, 0x1000);         // Disassemble one instruction at a time, offsetting addresses by 0x1000
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "lea ecx, dword ptr [edx + esi + 8]");
    range.popFront;
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "add eax, ebx");
    range.popFront;
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "add esi, 0x1234");
    range.popFront;
    assert(range.empty);
    ---
    */
    abstract InstructionRange disasmIter(in ubyte[] code, in ulong address) const;
}

// TODO: Try switching to InputRange!Instruction (more restrictive than isInputRange, though)
/// An input range that provides access to one disassembled `Instruction` at a time
abstract class InstructionRange {
    /// Retrieves element of the range
    Instruction front();
    /// True if range has no instructions, i.e. cannot be advanced anymore
    bool empty();
    /// Drops the front instruction and advances the range
    void popFront();
}
static assert(isInputRange!InstructionRange);

unittest{ // disasm
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = create(Arch.x86, ModeFlags(Mode.bit16));
    cs.mode = ModeFlags(Mode.bit32);

    auto res = cs.disasm(CODE, 0x1000);
    assert(res.length == 3); // With skipdata disabled, disassembling will halt when encountering data
    assert("%s %s".format(res[0].mnemonic, res[0].opStr) == "lea ecx, [edx + esi + 8]");
    assert("%s %s".format(res[1].mnemonic, res[1].opStr) == "add eax, ebx");
    assert("%s %s".format(res[2].mnemonic, res[2].opStr) == "add esi, 0x1234");
    cs.skipData = true;
    res = cs.disasm(CODE, 0x1000, 5);
    assert(res.length == 5);
}

/** Class template that encapsulates an architecture-specific instance of the Capstone dissassembly engine

Note that all architecture-specific instances, like `CapstoneX86`, instantiate and derive from this one.

Params:
    TInstructionId = The architecture-specific instruction identifier type
    TInstruction = The architecture-specific instruction type
*/
abstract class CapstoneImpl(TInstructionId, TInstruction) : Capstone {
    import capstone.range: InstructionImplRange;

    /** Creates an architecture-specific instance with a given mode of interpretation
    
    Params:
        arch = The architecture the engine will be created for
        modeFlags = The (initial) mode of interpretation, which can still be changed later on
    */
    package this(in Arch arch, in ModeFlags modeFlags){
        super(arch, modeFlags);
    }

    override TInstruction[] disasm(in ubyte[] code, in ulong address, in size_t count = 0) const {
        auto instrAppnd = appender!(TInstruction[]);
        foreach(i, instr; disasmIter(code, address).enumerate){
            instrAppnd.put(instr);
            if(i+1==count)
                break;
        }
        return instrAppnd.data;
    }

    override InstructionImplRange!TInstruction disasmIter(in ubyte[] code, in ulong address) const {
        return new InstructionImplRange!TInstruction(this, code, address);
    }

    /** Defines a custom mnemonic for a specified instruction id
    
    Example:
    ---
    enum X86_CODE32 = cast(ubyte[])"\x75\x01";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32));
    
    // Customize mnemonic JNE to JNZ
    cs.customMnemonic(X86InstructionId.jne, "jnz");
    
    foreach(instr; cs.disasm(X86_CODE32, 0x1000))
        writefln("%s\t%s", instr.mnemonic, instr.opStr);
    
    // Reset engine to use the default mnemonic of JNE
    cs.customMnemonic(X86InstructionId.jne);
    
    foreach(instr; cs.disasm(X86_CODE32, 0x1000))
        writefln("%s\t%s", instr.mnemonic, instr.opStr);
    ---
    */
    void customMnemonic(in TInstructionId id, in string mnem = null) {
        super.customMnemonic(id, mnem);
    }
}