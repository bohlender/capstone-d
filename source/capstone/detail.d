/// Object-oriented wrapper of Capstone's instruction detail
module capstone.detail;

import capstone.capstone: Capstone;
import capstone.internal: cs_detail;
import capstone.register;
import capstone.instructiongroup;

/// Instruction detail 
abstract class Detail {
private:
    const Capstone cs;
    cs_detail* internal;

public:
    /// Registers implicitly read by this instruction
	const(Register)[] regsRead() const;
    /// Registers implicitly modified by this instruction
	const(Register)[] regsWrite() const;
    /// The groups this instruction belongs to
	const(InstructionGroup)[] groups() const;

    private this(in Capstone cs, cs_detail* internal){
        this.cs = cs;
        this.internal = internal;
    }
}

/** Class template that encapsulates architecture-specific instruction detail

Note that all architecture-specific instances, like `X86Detail`, instantiate and derive from this one.
*/
abstract class DetailImpl(TRegister, TInstructionGroup, TInstructionDetail) : Detail {
    import std.array: array;
    import std.algorithm: map;

    /// Architecture-specific instruction detail
    TInstructionDetail archSpecific;
    /// Convenience-alias making `archSpecific`'s members directly accessible from this
    alias archSpecific this;

    package this(in Capstone cs, cs_detail* internal){
        super(cs, internal);
        archSpecific = TInstructionDetail(cs, internal.arch_detail);
    }

	override TRegister[] regsRead() const {return internal.regs_read[0..internal.regs_read_count].map!(reg => new TRegister(cs, reg)).array;}
	override TRegister[] regsWrite() const {return internal.regs_write[0..internal.regs_write_count].map!(reg => new TRegister(cs, reg)).array;}
	override TInstructionGroup[] groups() const {return internal.groups[0..internal.groups_count].map!(grp => new TInstructionGroup(cs, grp)).array;}
}