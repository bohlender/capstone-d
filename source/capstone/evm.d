/// Types and constants of EVM architecture
module capstone.evm;

import std.conv: to;

import capstone.api;
import capstone.capstone;
import capstone.detail;
import capstone.instruction;
import capstone.instructiongroup;
import capstone.internal;
import capstone.register;
import capstone.utils;

/** Architecture-specific Register variant

Note that this is a dummy enum as there are no valid registers
*/
class EvmRegister : RegisterImpl!EvmRegisterId {
    this(in Capstone cs, in int id) {
        super(cs, id);
    }
}

/// Architecture-specific InstructionGroup variant
class EvmInstructionGroup : InstructionGroupImpl!EvmInstructionGroupId {
    this(in Capstone cs, in int id) {
        super(cs, id);
    }
}

/// Architecture-specific Detail variant
class EvmDetail : DetailImpl!(EvmRegister, EvmInstructionGroup, EvmInstructionDetail) {
    this(in Capstone cs, cs_detail* internal) {
		super(cs, internal);
	}
}

/// Architecture-specific instruction variant
class EvmInstruction : InstructionImpl!(EvmInstructionId, EvmRegister, EvmDetail) {
    this(in Capstone cs, cs_insn* internal) {
		super(cs, internal);
	}
}

/// Architecture-specific Capstone variant
class CapstoneEvm : CapstoneImpl!(EvmInstructionId, EvmInstruction) {
    this(in ModeFlags modeFlags){
        super(Arch.evm, modeFlags);
    }
}

/// Evm-specific information about an instruction
struct EvmInstructionDetail {
    ubyte pop;  /// Number of items popped from the stack
    ubyte push; /// Number of items pushed into the stack
    uint fee;   /// Gas fee for the instruction

    package this(in Capstone cs, cs_arch_detail arch_detail){
        auto internal = arch_detail.evm;
		pop = internal.pop;
		push = internal.push;
		fee = internal.fee;
    }
}

//=============================================================================
// Constants
//=============================================================================

/** EVM register

Note that this is a dummy enum as there are no valid registers
*/
enum EvmRegisterId {
	invalid = 0
}

/// EVM instruction
enum EvmInstructionId {
	stop = 0,
	add = 1,
	mul = 2,
	sub = 3,
	div = 4,
	sdiv = 5,
	mod = 6,
	smod = 7,
	addmod = 8,
	mulmod = 9,
	exp = 10,
	signextend = 11,
	lt = 16,
	gt = 17,
	slt = 18,
	sgt = 19,
	eq = 20,
	iszero = 21,
	and = 22,
	or = 23,
	xor = 24,
	not = 25,
	byte_ = 26,
	sha3 = 32,
	address = 48,
	balance = 49,
	origin = 50,
	caller = 51,
	callvalue = 52,
	calldataload = 53,
	calldatasize = 54,
	calldatacopy = 55,
	codesize = 56,
	codecopy = 57,
	gasprice = 58,
	extcodesize = 59,
	extcodecopy = 60,
	returndatasize = 61,
	returndatacopy = 62,
	blockhash = 64,
	coinbase = 65,
	timestamp = 66,
	number = 67,
	difficulty = 68,
	gaslimit = 69,
	pop = 80,
	mload = 81,
	mstore = 82,
	mstore8 = 83,
	sload = 84,
	sstore = 85,
	jump = 86,
	jumpi = 87,
	pc = 88,
	msize = 89,
	gas = 90,
	jumpdest = 91,
	push1 = 96,
	push2 = 97,
	push3 = 98,
	push4 = 99,
	push5 = 100,
	push6 = 101,
	push7 = 102,
	push8 = 103,
	push9 = 104,
	push10 = 105,
	push11 = 106,
	push12 = 107,
	push13 = 108,
	push14 = 109,
	push15 = 110,
	push16 = 111,
	push17 = 112,
	push18 = 113,
	push19 = 114,
	push20 = 115,
	push21 = 116,
	push22 = 117,
	push23 = 118,
	push24 = 119,
	push25 = 120,
	push26 = 121,
	push27 = 122,
	push28 = 123,
	push29 = 124,
	push30 = 125,
	push31 = 126,
	push32 = 127,
	dup1 = 128,
	dup2 = 129,
	dup3 = 130,
	dup4 = 131,
	dup5 = 132,
	dup6 = 133,
	dup7 = 134,
	dup8 = 135,
	dup9 = 136,
	dup10 = 137,
	dup11 = 138,
	dup12 = 139,
	dup13 = 140,
	dup14 = 141,
	dup15 = 142,
	dup16 = 143,
	swap1 = 144,
	swap2 = 145,
	swap3 = 146,
	swap4 = 147,
	swap5 = 148,
	swap6 = 149,
	swap7 = 150,
	swap8 = 151,
	swap9 = 152,
	swap10 = 153,
	swap11 = 154,
	swap12 = 155,
	swap13 = 156,
	swap14 = 157,
	swap15 = 158,
	swap16 = 159,
	log0 = 160,
	log1 = 161,
	log2 = 162,
	log3 = 163,
	log4 = 164,
	create = 240,
	call = 241,
	callcode = 242,
	return_ = 243,
	delegatecall = 244,
	callblackbox = 245,
	staticcall = 250,
	revert = 253,
	suicide = 255,

	invalid = 512
}

/// Group of EVM instructions
enum EvmInstructionGroupId {
	invalid = 0,   // cs_grp_invalid

	jump,          // all jump instructions

	math = 8,      // math instructions
	stack_write,   // instructions write to stack
	stack_read,    // instructions read from stack
	mem_write,     // instructions write to memory
	mem_read,      // instructions read from memory
	store_write,   // instructions write to storage
	store_read,    // instructions read from storage
	halt   		   // instructions halt execution
}