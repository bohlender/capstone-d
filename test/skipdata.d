module test.skipdata;

import std.array: array;
import std.algorithm: filter;
import std.string: representation;
import std.outbuffer: OutBuffer;
import std.functional: toDelegate;

import capstone;
import test.utils;

enum X86_CODE32 = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
enum RANDOM_CODE = cast(ubyte[])"\xed\x00\x00\x00\x00\x1a\x5a\x0f\x1f\xff\xc2\x09\x80\x00\x00\x00\x07\xf7\xeb\x2a\xff\xff\x7f\x57\xe3\x01\xff\xff\x7f\x57\xeb\x00\xf0\x00\x00\x24\xb2\x4f\x00\x78";

//size_t delegate() mydelegate = (in ubyte[] code, size_t offset){
//	return 2UL;
//}

size_t mycallback(in ubyte[] code, size_t offset) nothrow @nogc{
    return 2; // Always skip 2 bytes when encountering data
}

enum platforms = [
    Platform(Arch.x86, Mode.bit32, X86_CODE32, "X86 32 (Intel syntax) - Skip data", Syntax.systemDefault),
    Platform(Arch.arm, Mode.arm, RANDOM_CODE, "Arm - Skip data", Syntax.systemDefault),
    Platform(Arch.x86, Mode.bit32, X86_CODE32, "X86 32 (Intel syntax) - Skip data with custom mnemonic", Syntax.systemDefault, "db"),
    Platform(Arch.arm, Mode.arm, RANDOM_CODE, "Arm - Skip data with callback", Syntax.systemDefault, "db")
];

unittest{
    // TODO: Use compile-time delegate instead once dmd 2.0.83 is stable (see: https://issues.dlang.org/show_bug.cgi?id=17899)
	Callback[platforms.length] callbacks;
    callbacks[3] = toDelegate(&mycallback);

    auto buf = new OutBuffer;
    static foreach(i, platform; platforms) {{
        auto cs = new Capstone!(platform.arch)(ModeFlags(platform.mode));
        cs.syntax = platform.syntax;
        cs.detail = true;
        cs.skipData = true;

        if(platform.skipdataMnemonic)
            cs.setupSkipdata(platform.skipdataMnemonic, callbacks[i]);
        
        auto res = cs.disasm(platform.code, 0x1000);
        assert(res.length > 0);

        buf.writefln("****************");
        buf.writefln("Platform: %s", platform.comment);
        buf.writefln("Code: %s", platform.code.bytesToHex);
        buf.writefln("Disasm:");

        foreach(instr; res){
            buf.writefln("0x%x:\t%s\t\t%s", instr.address, instr.mnemonic, instr.opStr);
        }
        buf.writefln("0x%x:", res[$-1].address + res[$-1].bytes.length);
        buf.writefln("");
    }}

    const expected = import("skipdata.expected");
    const actual = buf.toString;

    assert(expected == actual, expectationMismatch(expected, actual));
}
