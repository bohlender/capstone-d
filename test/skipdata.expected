****************
Platform: X86 32 (Intel syntax) - Skip data
Code: 0x8d 0x4c 0x32 0x08 0x01 0xd8 0x81 0xc6 0x34 0x12 0x00 0x00 0x00 0x91 0x92 
Disasm:
0x1000:	lea		ecx, [edx + esi + 8]
0x1004:	add		eax, ebx
0x1006:	add		esi, 0x1234
0x100c:	.byte		0x00
0x100d:	xchg		eax, ecx
0x100e:	xchg		eax, edx
0x100f:

****************
Platform: X86 32 (Intel syntax) - Skip data with custom mnemonic
Code: 0x8d 0x4c 0x32 0x08 0x01 0xd8 0x81 0xc6 0x34 0x12 0x00 0x00 0x00 0x91 0x92 
Disasm:
0x1000:	lea		ecx, [edx + esi + 8]
0x1004:	add		eax, ebx
0x1006:	add		esi, 0x1234
0x100c:	db		0x00
0x100d:	xchg		eax, ecx
0x100e:	xchg		eax, edx
0x100f:

****************
Platform: Arm - Skip data
Code: 0xed 0x00 0x00 0x00 0x00 0x1a 0x5a 0x0f 0x1f 0xff 0xc2 0x09 0x80 0x00 0x00 0x00 0x07 0xf7 0xeb 0x2a 0xff 0xff 0x7f 0x57 0xe3 0x01 0xff 0xff 0x7f 0x57 0xeb 0x00 0xf0 0x00 0x00 0x24 0xb2 0x4f 0x00 0x78 
Disasm:
0x1000:	andeq		r0, r0, sp, ror #1
0x1004:	svceq		#0x5a1a00
0x1008:	stmibeq		r2, {r0, r1, r2, r3, r4, r8, sb, sl, fp, ip, sp, lr, pc} ^
0x100c:	andeq		r0, r0, r0, lsl #1
0x1010:	bhs		#0xffafec34
0x1014:	.byte		0xff, 0xff, 0x7f, 0x57
0x1018:	.byte		0xe3, 0x01, 0xff, 0xff
0x101c:	rsceq		r5, fp, pc, ror r7
0x1020:	strhs		r0, [r0], #-0xf0
0x1024:	stmdavc		r0, {r1, r4, r5, r7, r8, sb, sl, fp, lr}
0x1028:

****************
Platform: Arm - Skip data with callback
Code: 0xed 0x00 0x00 0x00 0x00 0x1a 0x5a 0x0f 0x1f 0xff 0xc2 0x09 0x80 0x00 0x00 0x00 0x07 0xf7 0xeb 0x2a 0xff 0xff 0x7f 0x57 0xe3 0x01 0xff 0xff 0x7f 0x57 0xeb 0x00 0xf0 0x00 0x00 0x24 0xb2 0x4f 0x00 0x78 
Disasm:
0x1000:	andeq		r0, r0, sp, ror #1
0x1004:	svceq		#0x5a1a00
0x1008:	stmibeq		r2, {r0, r1, r2, r3, r4, r8, sb, sl, fp, ip, sp, lr, pc} ^
0x100c:	andeq		r0, r0, r0, lsl #1
0x1010:	bhs		#0xffafec34
0x1014:	db		0xff, 0xff
0x1016:	db		0x7f, 0x57
0x1018:	db		0xe3, 0x01
0x101a:	db		0xff, 0xff
0x101c:	rsceq		r5, fp, pc, ror r7
0x1020:	strhs		r0, [r0], #-0xf0
0x1024:	stmdavc		r0, {r1, r4, r5, r7, r8, sb, sl, fp, lr}
0x1028:

